archeryutils/handicaps/handicap_scheme.py

```diff
@@ -0,0 +1,764 @@
+        score : FloatArray
+            raw scores to be rounded according to the handicap system convention
+
+        Returns
+        -------
+        FloatArray
+            scores after appropriate rounding
+        """
+        return np.around(score)
+
+    def handicap_from_score(
+        self,
+        score: float,
+        rnd: rounds.Round,
+        arw_d: Optional[float] = None,
+        int_prec: bool = False,
+    ) -> Union[int, float]:
+        """Calculate the handicap of a given score on a given round.
+
+        Parameters
+        ----------
+        score : float
+            score achieved on the round
+        rnd : rounds.Round
+            the rounds.Round object to calculate the handicap for
+        arw_d : float or None, default=None
+            user-specified arrow diameter in [metres]
+        int_prec : bool, default=False
+            display results as integers? default = False
+            decimal results accurate to 2dp from rootfinder
+
+        Returns
+        -------
+        handicap: int or float
+            Handicap for score. Has type int if int_prec is True, else float.
+
+        Raises
+        ------
+        ValueError
+            If an invalid score for the given round is provided.
+
+        Examples
+        --------
+        Handicap for a score of 999 on a WA 1440 (90m),
+        using the AGB handicap system (via the HandicapSchemeAGB subclass),
+        can be calculated using:
+
+        >>> import archeryutils as au
+        >>> import archeryutils.handicaps as hc
+        >>> wa_outdoor = au.load_rounds.WA_outdoor
+        >>> agb_scheme = hc.handicap_scheme("AGB")
+        >>> agb_scheme.score_for_round(wa_outdoor.wa1440_90, 10.0)
+        >>> agb_scheme.handicap_from_score(999, wa_outdoor.wa1440_90)
+        43.999964586102706
+
+        To get an integer value as would appear in the handicap tables use
+        ``int_prec=True``:
+
+        >>> agb_scheme.handicap_from_score(999, wa_outdoor.wa1440_90), int_prec=True)
+        44.0
+
+        """
+        # Check we have a valid score
+        max_score = rnd.max_score()
+        if score > max_score:
+            raise ValueError(
+                f"The score of {score} provided is greater than the maximum of {max_score} "
+                f"for a {rnd.name}."
+            )
+        if score <= 0.0:
+            raise ValueError(
+                f"The score of {score} provided is less than or equal to zero so cannot "
+                "have a handicap."
+            )
+
+        if score == max_score:
+            # Deal with max score before root finding
+            return self._get_max_score_handicap(rnd, arw_d, int_prec)
+
+        handicap = self._rootfind_score_handicap(score, rnd, arw_d=arw_d)
+
+        # Force integer precision if required.
+        if int_prec:
+            if self.desc_scale:
+                handicap = np.ceil(handicap)
+            else:
+                handicap = np.floor(handicap)
```

Comment:
Same point on coupling of `desc_scale` to implementation specific AA/AGB rules. I think desc_scale should only be used to influence searching upwards or downards on the handicap scale in these algorithms, but should definitly avoid keying other behaviour off it.

Reply:
I think this is OK, as the decision on whether you round to the integer above or below is dependent on whether the handicap scale is ascending or descending.

Reply:
Yeah, on re-reading I twigged that this time, any scheme with an ascending/descending scale would need to choose ceil or floor appropraitely